db/db_impl.cc

```diff
@@ -630,7 +634,96 @@ void DBImpl::StartTimedTasks() {
+      if (!thread_persist_stats_) {
+        thread_persist_stats_.reset(new rocksdb::RepeatableThread(
+            [this]() { DBImpl::PersistStats(); }, "pst_st", env_,
+            stats_persist_period_sec * 1000000));
+      }
+    }
+  }
+}
+
+void DBImpl::PersistStats() {
+  TEST_SYNC_POINT("DBImpl::PersistStats:1");
+#ifndef ROCKSDB_LITE
+  const DBPropertyInfo* cf_property_info =
+      GetPropertyInfo(DB::Properties::kCFStats);
+  assert(cf_property_info != nullptr);
+  const DBPropertyInfo* db_property_info =
+      GetPropertyInfo(DB::Properties::kDBStats);
+  assert(db_property_info != nullptr);
+  std::map<std::string, std::string> stats_map;
+  if (shutdown_initiated_) {
+    return;
+  }
+  {
+    InstrumentedMutexLock l(&mutex_);
+    default_cf_internal_stats_->GetMapProperty(
+        *db_property_info, DB::Properties::kDBStats, &stats_map);
+    std::map<std::string, std::string> cf_stats_map;
+    for (auto cfd : *versions_->GetColumnFamilySet()) {
+      if (cfd->initialized()) {
+        cfd->internal_stats()->GetMapProperty(
+            *cf_property_info, DB::Properties::kCFStatsNoFileHistogram,
+            &cf_stats_map);
+        cfd->internal_stats()->GetMapProperty(
+            *cf_property_info, DB::Properties::kCFFileHistogram, &cf_stats_map);
+        // add column family name to the key string
+        for (const auto& cf_stats : cf_stats_map) {
+          std::string key_with_cf_name = cfd->GetName() + "." + cf_stats.first;
+          stats_map[key_with_cf_name] = cf_stats.second;
+        }
+      }
+    }
+    TEST_SYNC_POINT("DBImpl::PersistStats:2");
+    WriteOptions wo;
+    ColumnFamilyOptions cf_options;
+    const uint64_t now_micros = env_->NowMicros();
+    stats_history_[now_micros] = stats_map;
+    // delete older stats snapshots to control memory consumption
+    int num_stats_to_delete =
+        static_cast<int>(stats_history_.size()) -
+        static_cast<int>(mutable_db_options_.max_stats_history_count);
+    if (num_stats_to_delete > 0) {
+      auto it = stats_history_.begin();
+      for (int i = 0; i < num_stats_to_delete; ++i) {
+        if (it != stats_history_.end()) {
+          it = stats_history_.erase(it);
+        }
+      }
+    }
+    // TODO: also persist stats to disk
   }
+#endif  // !ROCKSDB_LITE
+}
+
+bool DBImpl::FindStatsByTime(uint64_t start_time, uint64_t end_time,
+                             uint64_t* new_time,
+                             std::map<std::string, std::string>* stats_map) {
+  // lock when search for start_time
+  {
+    InstrumentedMutexLock l(&mutex_);
+    auto it = stats_history_.lower_bound(start_time);
+    if (it != stats_history_.end() && it->first < end_time) {
+      // make a copy for timestamp and stats_map
+      *new_time = it->first;
+      *stats_map = it->second;
+      return true;
+    }
+    else {
+      return false;
+    }
+  }
+}
+
+Status DBImpl::GetStatsHistory(GetStatsOptions& stats_opts,
```

Comment:
I would argue that start and end time should be argument, not an option. It's just like in Get(), key is not an option.